actors/evm/src/interpreter/precompiles/mod.rs

```diff
@@ -14,79 +14,101 @@ use fvm::{call_actor, get_actor_type, get_randomness, lookup_delegated_address,
-        lookup_delegated_address,     // lookup_delegated_address 0x0b
-        get_actor_type,     // get actor type 0x0c
-        get_randomness,     // rand 0x0d
-        call_actor,         // call_actor 0x0e
+        ec_recover, // 0x01 ecrecover
+        sha256,     // 0x02 SHA2-256
+        ripemd160,  // 0x03 ripemd160
+        identity,   // 0x04 identity
+        modexp,     // 0x05 modexp
+        ec_add,     // 0x06 ecAdd
+        ec_mul,     // 0x07 ecMul
+        ec_pairing, // 0x08 ecPairing
+        blake2f,    // 0x09 blake2f
+    }
+}
+
+const fn gen_native_precompiles<RT: Runtime>() -> [PrecompileFn<RT>; 5] {
+    precompiles! {
+        resolve_address,            // 0xfe00..01 resolve_address
+        lookup_delegated_address,   // 0xfe00..02 lookup_delegated_address
+        call_actor,                 // 0xfe00..03 call_actor
+        get_actor_type,             // 0xfe00..04 get_actor_type
+        get_randomness,             // 0xfe00..05 rand
     }
 }
 
 pub struct Precompiles<RT>(PhantomData<RT>);
 
 impl<RT: Runtime> Precompiles<RT> {
-    const PRECOMPILES: [PrecompileFn<RT>; 14] = gen_precompiles();
-    const MAX_PRECOMPILE: U256 = {
-        let mut limbs = [0u64; 4];
-        limbs[0] = Self::PRECOMPILES.len() as u64;
-        U256(limbs)
-    };
-
-    // Precompile Context will be flattened to None if not calling the call_actor precompile
+    const EVM_PRECOMPILES: [PrecompileFn<RT>; 9] = gen_evm_precompiles();
+    const NATIVE_PRECOMPILES: [PrecompileFn<RT>; 5] = gen_native_precompiles();
+
+    fn lookup_precompile(addr: &[u8; 32]) -> Option<PrecompileFn<RT>> {
+        // unrwap will never panic, 32 - 12 = 20
+        let addr: [u8; 20] = addr[12..].try_into().unwrap();
+        let [prefix, middle @ .., index] = addr;
+        if middle == [0u8; 18] && index > 0 {
+            let index = index as usize - 1;
+            match prefix {
+                NATIVE_PRECOMPILE_ADDRESS_PREFIX => Some(Self::NATIVE_PRECOMPILES[index]),
+                0x00 => Some(Self::EVM_PRECOMPILES[index]),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Precompile Context will be flattened to None if not calling the call_actor precompile.
+    /// Panics if address is not a precompile.
     pub fn call_precompile(
         system: &mut System<RT>,
         precompile_addr: U256,
         input: &[u8],
         context: PrecompileContext,
     ) -> PrecompileResult {
-        unsafe { Self::PRECOMPILES[precompile_addr.0[0] as usize - 1](system, input, context) }
+        unsafe {
+            Self::lookup_precompile(&precompile_addr.to_bytes()).unwrap()(system, input, context)
+        }
     }
 
+    /// Checks if word represents
     #[inline]
     pub fn is_precompile(addr: &U256) -> bool {
-        !addr.is_zero() && addr <= &Self::MAX_PRECOMPILE
+        !addr.is_zero() && Self::lookup_precompile(&addr.to_bytes()).is_some()
```

Comment:
We can probably drop the "is zero" check (or we can just leave it, doesn't really matter).